### PR TITLE
Feature/doo hickey.v2

### PR DESCRIPTION
**Pull request by jasonplatte**
PTAL

### CODE REVIEW

File: src/main/cpp/subsystems/DooHickey.cpp [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+constexpr frc::DoubleSolenoid::Value kHickeyEngage {frc::DoubleSolenoid::kForward};
+constexpr frc::DoubleSolenoid::Value kHickeyDisengage {frc::DoubleSolenoid::kReverse};
+
+DooHickey::DooHickey(
+    const wpi::Twine& name,
+    WPI_TalonFX& spinner,
+    frc::DoubleSolenoid& UpPushyThang)
+  : spinner(0)
+  ,UpPushyThang(0,7)
+  ,mLogfile("/home/lvuser/test.txt", std::ios_base::out)
```
Comment (duane701) [inline src/main/cpp/subsystems/DooHickey.cpp]: We can remove all of the log file stuff. The log file has served its purpose.

File: src/main/cpp/subsystems/DooHickey.cpp [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    spinner.ConfigNominalOutputReverse(0, 10);
+    spinner.ConfigPeakOutputForward(1.0, 10);
+    spinner.ConfigPeakOutputReverse(-1.0, 10);
+
+    /* Set Motion Magic gains in slot0 - see documentation */
+    spinner.SelectProfileSlot(0, 0);
+    spinner.Config_kF(0, 0.3, 10);
+    spinner.Config_kP(0, 0.1, 10);
+    spinner.Config_kI(0, 0.0, 10);
+    spinner.Config_kD(0, 0.0, 10);
```
Comment (duane701) [inline src/main/cpp/subsystems/DooHickey.cpp]: Let's make P, I, D, F constants in this file that will be easier for use to find and change.

For example, near the beginning of this file

namespace
{
  constexpr double kF{0.3};
  constexpr double kP{0.1};
  // etc for I and D
}

File: src/main/cpp/subsystems/DooHickey.cpp [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+}
+
+void DooHickey::Init() {
+    
+    m_colorMatcher.AddColorMatch(kBlueTarget);
+    m_colorMatcher.AddColorMatch(kGreenTarget);
+    m_colorMatcher.AddColorMatch(kRedTarget);
+    m_colorMatcher.AddColorMatch(kYellowTarget);
+
+    mPreviousTime = std::chrono::system_clock::now();
```
Comment (duane701) [inline src/main/cpp/subsystems/DooHickey.cpp]: Let's remove all the timing stuff as well.

File: src/main/cpp/subsystems/DooHickey.cpp [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+#include "subsystems/DooHickey.h"
+#include <ctime>
+#include <fstream>
+#include <sstream>
+
+constexpr frc::DoubleSolenoid::Value kHickeyEngage {frc::DoubleSolenoid::kForward};
+constexpr frc::DoubleSolenoid::Value kHickeyDisengage {frc::DoubleSolenoid::kReverse};
```
Comment (duane701) [inline src/main/cpp/subsystems/DooHickey.cpp]: Put these two lines inside of `namespace {`.

File: src/main/include/subsystems/DooHickey.h [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+#pragma once
+
+#include <ctre/phoenix/motorcontrol/can/WPI_TalonFX.h>
+#include <frc2/command/SubsystemBase.h>
+#include <chrono>
+#include <fstream>
```
Comment (duane701) [inline src/main/include/subsystems/DooHickey.h]: Remove `chrono` and `fstream`

File: src/main/include/subsystems/DooHickey.h [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+#include <frc2/command/button/JoystickButton.h>
+#include <frc/util/color.h>
+#include <frc/WPILib.h>
+#include "ctre/Phoenix.h"
+#include <frc/DoubleSolenoid.h>
```
Comment (duane701) [inline src/main/include/subsystems/DooHickey.h:14]: Remove `WPILIb.h` and `Phoenix.h`. The includes here are close. Add any additional includes for pieces that are missing.

File: src/main/include/subsystems/DooHickey.h [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+  void MoveSpinner(double speed);
+
+
+ private:
+ 
+  WPI_TalonFX& spinner;
+  frc::DoubleSolenoid& UpPushyThang;
+
+  std::ofstream mLogfile;
+  time_point mPreviousTime;
```
Comment (duane701) [inline src/main/include/subsystems/DooHickey.h]: Remove mPreviousTime and mLogFile.

File: src/main/cpp/RobotContainer.cpp [hunk 8]
```diff
@@ -21,7 +24,14 @@ RobotContainer::RobotContainer()
       mChassis
     )
   );
+  double radiusCW = 16; //16in radius 
+  double radiusDW = 2;  //change this later
 
+  frc::SmartDashboard::PutData("Spin 600 RPM", new Spin(mDooHickey, 0.1));
+  frc::SmartDashboard::PutData("Spin 6000 RPM", new Spin(mDooHickey, 0.9404));
```
Comment (duane701) [inline src/main/cpp/RobotContainer.cpp]: Nice random numbers.

File: src/main/cpp/commands/SetHickeyPos.cpp [hunk 9]
```diff
@@ -0,0 +1,21 @@
+/*----------------------------------------------------------------------------*/
+/* Copyright (c) 2019 FIRST. All Rights Reserved.                             */
+/* Open Source Software - may be modified and shared by FRC teams. The code   */
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+#include "commands/SetHickeyPos.h"
+
+SetHickeyPos::SetHickeyPos(DooHickey* dooHickey, double position) : mDooHickey(dooHickey), mPosition(position) {
```
Comment (duane701) [inline src/main/cpp/commands/SetHickeyPos.cpp]: Let's use a reference to the subsystem.

```suggestion
SetHickeyPos::SetHickeyPos(DooHickey& dooHickey, double position) : mDooHickey(dooHickey), mPosition(position) {
```

File: src/main/include/commands/SetHickeyPos.h [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+#include <frc2/command/CommandHelper.h>
+#include <frc2/command/InstantCommand.h>
+#include "subsystems/DooHickey.h"
+
+class SetHickeyPos
+    : public frc2::CommandHelper<frc2::InstantCommand,
+                                 SetHickeyPos> {
+ public:
+  SetHickeyPos(DooHickey* dooHickey, double position);
```
Comment (duane701) [inline src/main/include/commands/SetHickeyPos.h]: Use a reference for `dooHickey`

```suggestion
  SetHickeyPos(DooHickey& dooHickey, double position);
```

File: src/main/include/commands/SetHickeyPos.h [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+
+  void Initialize() override;
+ private:
+  DooHickey* mDooHickey;
```
Comment (duane701) [inline src/main/include/commands/SetHickeyPos.h]: Reference

```suggestion
  DooHickey& mDooHickey;
```

File: src/main/include/commands/HickeyEngage.h [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+
+class HickeyEngage
+    : public frc2::CommandHelper<frc2::InstantCommand,
+                                 HickeyEngage> {
+ public:
+  HickeyEngage();
+
+  void Initialize() override;
+ private:
+  DooHickey* mDooHickey;
```
Comment (duane701) [inline src/main/include/commands/HickeyEngage.h]: `mDooHickey` is uninitialized and use a reference.

```suggestion
  DooHickey& mDooHickey;
```

File: src/main/include/commands/HickeyEngage.h [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+
+#include <frc2/command/CommandHelper.h>
+#include <frc2/command/InstantCommand.h>
+#include "subsystems/DooHickey.h"
+
+class HickeyEngage
+    : public frc2::CommandHelper<frc2::InstantCommand,
+                                 HickeyEngage> {
+ public:
+  HickeyEngage();
```
Comment (duane701) [inline src/main/include/commands/HickeyEngage.h]: Need a subsystem

```suggestion
  HickeyEngage(DooHickey& dooHickey);
```

File: src/main/include/commands/HickeyDisengage.h [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+
+#include <frc2/command/CommandHelper.h>
+#include <frc2/command/InstantCommand.h>
+#include "subsystems/DooHickey.h"
+
+class HickeyDisengage
+    : public frc2::CommandHelper<frc2::InstantCommand,
+                                 HickeyDisengage> {
+ public:
+  HickeyDisengage();
```
Comment (duane701) [inline src/main/include/commands/HickeyDisengage.h]: Need a subsystem.

File: src/main/include/commands/HickeyDisengage.h [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+
+  void Initialize() override;
+ private:
+  DooHickey* mDooHickey;
```
Comment (duane701) [inline src/main/include/commands/HickeyDisengage.h]: mDooHickey is uninitialized and should be a reference.

```suggestion
  DooHickey& mDooHickey;
```

File: src/main/cpp/commands/HickeyEngage.cpp [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+#include "commands/HickeyEngage.h"
+
+// NOTE:  Consider using this command inline, rather than writing a subclass.
+// For more information, see:
+// https://docs.wpilib.org/en/latest/docs/software/commandbased/convenience-features.html
+HickeyEngage::HickeyEngage() {
```
Comment (duane701) [inline src/main/cpp/commands/HickeyEngage.cpp]: Need a subsystem as a reference.

File: src/main/cpp/commands/HickeyEngage.cpp [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+  SetName("HickeyEngage");
+  AddRequirements(mDooHickey);
```
Comment (duane701) [inline src/main/cpp/commands/HickeyEngage.cpp]: When `mDooHickey` is a reference, change this too

```suggestion
  AddRequirements(&mDooHickey);
```

File: src/main/cpp/commands/HickeyDisengage.cpp [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+#include "commands/HickeyDisengage.h"
+
+// NOTE:  Consider using this command inline, rather than writing a subclass.
+// For more information, see:
+// https://docs.wpilib.org/en/latest/docs/software/commandbased/convenience-features.html
+HickeyDisengage::HickeyDisengage() {
```
Comment (duane701) [inline src/main/cpp/commands/HickeyDisengage.cpp]: Need subsystem as reference.

File: src/main/cpp/subsystems/DooHickey.cpp [hunk 19]
```diff
@@ -120,6 +125,68 @@ void DooHickey::Periodic() {
...
+    /*Tracks the number of times a specific color is detected (in this case blue), so that we can 
+    * know when rotation control is over. We should see each color 2 times a rotation and the goal is 3
+    * rotations.
+    */ 
+    if (colorCounter >= 6) {
+      IsRotationControlFinished = true;
+    }
+
+    //Code for stopping the motor after Rotation Control is finished AND the specified Color is reached.
+    gameData = frc::DriverStation::GetInstance().GetGameSpecificMessage();
```
Comment (duane701) [inline src/main/cpp/subsystems/DooHickey.cpp:140]: I think this is going to be moved. I think the message only comes through when it's time to spin the wheel.

File: src/main/include/Robot.h [hunk 20]
```diff
@@ -6,10 +6,10 @@
 /*----------------------------------------------------------------------------*/
 
 #pragma once
-
+#include <frc/DriverStation.h>
 #include <frc/TimedRobot.h>
 #include <frc2/command/Command.h>
-
+#include <subsystems/DooHickey.h>
```
Comment (duane701) [inline src/main/include/Robot.h]: Use `"` for our includes

```suggestion
#include "subsystems/DooHickey.h"
```

File: src/main/cpp/RobotContainer.cpp [hunk 21]
```diff
@@ -24,8 +24,8 @@ RobotContainer::RobotContainer()
       mChassis
     )
   );
-  double radiusCW = 16; //16in radius 
-  double radiusDW = 2;  //change this later
+  double radiusCW = 16; //16" radius of Control panel 
+  double radiusDW = 1.5;  //1.5" radius of DooHickey wheel (3" diameter)
```
Comment (duane701) [inline src/main/cpp/RobotContainer.cpp]: Good info. We'll probably need this somewhere else.

**Review by duane701 — CHANGES_REQUESTED** [review]

**Comment by duane701** [comment]
FTP

File: src/main/cpp/subsystems/DooHickey.cpp (hunk 4 above)
Comment (duane701) [reply src/main/cpp/subsystems/DooHickey.cpp]: And then add PIDF with them

**Comment by duane701** [comment]
LGTM -- Rebase after Kenyon 

Or you can race. First one wins. Second one has to rebase -- again.

**Comment by jasonplatte** [comment]
Replaced by #31

**Pull request closed without merging**